python/EBS_SNAPSHOT_PUBLIC_RESTORABLE_CHECK/EBS_SNAPSHOT_PUBLIC_RESTORABLE_CHECK.py

```diff
@@ -0,0 +1,388 @@
+#
+#        http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for
+# the specific language governing permissions and limitations under the License.
+'''
+#####################################
+##           Gherkin               ##
+#####################################
+Rule Name:
+  EBS_SNAPSHOT_PUBLIC_RESTORABLE_CHECK
+Description:
+  Check whether the Amazon EBS snapshots are not publicly restorable. The rule is NON_COMPLIANT if the RestorableByUserIds field is set to 'all'.
+Trigger:
+  Periodic
+Reports on:
+  AWS::::Account
+Rule Parameters:
+  None
+Scenarios:
+  Scenario: 1
+    Given: No snapshots with RestorableByUserIds parameter set to 'all'
+     Then: Return COMPLIANT
+  Scenario: 2
+    Given: One or more snapshots with RestorableByUserIds parameter set to 'all'
+     Then: Return NON_COMPLIANT with Annotation containing SnapshotIDs
+'''
+import json
+import datetime
+import boto3
+import botocore
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::::Account'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+#############
+# Main Code #
+#############
+
+
+# Function to generate evaluations for all non compliant resources
+def generate_snapshot_id_list(snapshots, event):
+    snapshot_ids = []
+    for snapshot in snapshots:
+        snapshot_ids.append(snapshot['SnapshotId'])
+    return snapshot_ids
+
+
+# Function to obtain all public Amazon EBS snapshots
+def get_public_snapshots(ec2_client, owner_id):
+    snapshots = []
+    next_token = None
+    snapshots_result = {}
+    while True:
+        if not next_token:
+            snapshots_result = ec2_client.describe_snapshots(Filters=[{'Name':'owner-id', 'Values':[owner_id]}], OwnerIds=[owner_id], RestorableByUserIds=['all'], MaxResults=1000)
+        else:
+            snapshots_result = ec2_client.describe_snapshots(Filters=[{'Name':'owner-id', 'Values':[owner_id]}], OwnerIds=[owner_id], RestorableByUserIds=['all'], NextToken=next_token, MaxResults=1000)
+        snapshots.extend(snapshots_result['Snapshots'])
+        if 'NextToken' in snapshots_result:
+            next_token = snapshots_result['NextToken']
+        else:
+            return(True, snapshots)
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    ec2_client = get_client('ec2', event)
+    public_snapshots_result = get_public_snapshots(ec2_client, event['accountId'])
+    snapshot_ids = generate_snapshot_id_list(public_snapshots_result[1], event)
+    if snapshot_ids:
+        return build_evaluation(event['accountId'], 'NON_COMPLIANT', event, annotation='Public Amazon EBS Snapshots: {}'.format(snapshot_ids))

```

Comment:
What happens when the annotation excess 256 characters?
https://docs.aws.amazon.com/config/latest/APIReference/API_Evaluation.html